components/script/textinput.rs

```diff
@@ -1172,81 +1180,103 @@ impl<T: ClipboardProvider> TextInput<T> {
 
-            // Step 3.2 Fire a clipboard event named clipboardchange
-            owning_node
-                .owner_document()
-                .fire_clipboardchange_event(can_gc);
-        },
-        "cut" => {
-            let selection = textinput.borrow().get_selection_text();
+                // Step 3.1 If there is a selection in an editable context where cutting is enabled, then
+                let Some(text) = selection else {
+                    // Step 3.2 Else, if there is no selection or the context is not editable, then
+                    return ClipboardEventReaction::empty();
+                };
 
-            // Step 3.1 If there is a selection in an editable context where cutting is enabled, then
-            if let Some(text) = selection {
                 // Step 3.1.1 Copy the selected contents, if any, to the clipboard
-                textinput.borrow_mut().clipboard_provider.set_text(text);
+                self.clipboard_provider.set_text(text);
 
                 // Step 3.1.2 Remove the contents of the selection from the document and collapse the selection.
-                textinput.borrow_mut().delete_char(Direction::Backward);
+                self.delete_char(Direction::Backward);
 
                 // Step 3.1.3 Fire a clipboard event named clipboardchange
-                owning_node
-                    .owner_document()
-                    .fire_clipboardchange_event(can_gc);
-
                 // Step 3.1.4 Queue tasks to fire any events that should fire due to the modification.
-            } else {
-                // Step 3.2 Else, if there is no selection or the context is not editable, then
-                return false;
-            }
-        },
-        "paste" => {
-            // Step 3.1 If there is a selection or cursor in an editable context where pasting is enabled, then
-            if let Some(data) = event.get_clipboard_data() {
-                // Step 3.1.1 Insert the most suitable content found on the clipboard, if any, into the context.
-                let drag_data_store = data.data_store().expect("This shouldn't fail");
-                textinput.borrow_mut().paste_contents(&drag_data_store);
-
-                // Step 3.1.2 Queue tasks to fire any events that should fire due to the modification.
-            } else {
-                // Step 3.2 Else return false.
-                return false;
-            }
-        },
-        _ => (),
-    }
+                ClipboardEventReaction::FireClipboardChangedEvent |
+                    ClipboardEventReaction::QueueInputEvent
+            },
+            "paste" => {
+                // These steps are from <https://www.w3.org/TR/clipboard-apis/#paste-action>:
+                let Some(data_transfer) = clipboard_event.get_clipboard_data() else {
+                    return ClipboardEventReaction::empty();
+                };
+                let Some(drag_data_store) = data_transfer.data_store() else {
```

Comment:
It seems that our `DataTransfer` struct was implemented with only drag-and-drop in mind...